packages/compass-data-modeling/src/components/diagram-editor.tsx

```diff
@@ -333,21 +367,121 @@ const DiagramContent: React.FunctionComponent<{
+      // since unlike react-flow, we only support single item selections
+      if (nodes.length > 0) {
+        onNodeClick(null, nodes[0]);
+        return;
+      }
+      if (edges.length > 0) {
+        onEdgeClick(null, edges[0]);
+        return;
+      }
+      // If nothing is selected, clear the selection
+      onDiagramBackgroundClicked(true);
+    },
+    [onDiagramBackgroundClicked, onNodeClick, onEdgeClick]
+  );
+
+  const onKeyDown = useCallback(
+    (event: React.KeyboardEvent | KeyboardEvent) => {
+      // If no element is focused, some inputs should still be treated as targeting the diagram
+      // (e.g. undo/redo)
+      const isUnfocusedEvent =
+        event.target === document.body ||
+        event.target === document.documentElement;
+      if (
+        !isUnfocusedEvent &&
+        !anyParentNodeIsDataModelingDiagramOrDiagramToolbar(
+          event.target as HTMLElement
+        )
+      ) {
+        return;
+      }
+      const markHandled = () => {
+        event.preventDefault();
+        event.stopPropagation();
+      };
+
+      // Backspace and Delete keys delete selected item
+      if (
+        !isUnfocusedEvent &&
+        (event.key === 'Backspace' || event.key === 'Delete')
+      ) {
+        markHandled();
+        switch (selectedItems?.type) {
+          case 'collection':
+            onDeleteCollection(selectedItems.id);
+            break;
+          case 'relationship':
+            onDeleteRelationship(selectedItems.id);
+            break;
+          case 'field':
+            onDeleteField(selectedItems.namespace, selectedItems.fieldPath);
+            break;
+          default:
+            break;
+        }
+      }
+
+      // Escape key clears selection
+      else if (!isUnfocusedEvent && event.key === 'Escape') {
+        markHandled();
+        onDiagramBackgroundClicked(true);
+        // Undo/Redo should also trigger if there is no currently focused element
+        // macOS: Cmd+Shift+Z = Redo, Cmd+Z = Undo
+      } else if (event.metaKey && event.key.toLowerCase() === 'z') {
+        markHandled();
+        if (event.shiftKey) {
+          onRedoClick();
+        } else {
+          onUndoClick();
+        }
+        // Windows/Linux: Ctrl+Z = Undo, Ctrl+Y = Redo
+      } else if (event.ctrlKey && event.key.toLowerCase() === 'z') {
+        markHandled();
+        onUndoClick();
+      } else if (event.ctrlKey && event.key.toLowerCase() === 'y') {
+        markHandled();
+        onRedoClick();
+      }
+    },
+    [
+      onDiagramBackgroundClicked,
+      onDeleteCollection,
+      onDeleteRelationship,
+      onDeleteField,
+      selectedItems,
+      onUndoClick,
+      onRedoClick,
+    ]
+  );
+
+  // Not sure why eslint is confused by `DiagramProps` here
+  // eslint-disable-next-line @typescript-eslint/no-redundant-type-constituents
+  type ThrottledDiagramProps = DiagramProps & {
+    onKeyDown: (event: React.KeyboardEvent | KeyboardEvent) => void;
```

Comment:
Why do we need the keydown here? As far as I understand we only need to throttle props that are being passed to diagramming package and as this is a somewhat hacky workaround for the library performance issues, we should probably not add more stuff here unless it is really needed

Reply:
Yep, this is gone away with the switch to useHotkeys anyway 🙂